src/charm.py

```diff
@@ -38,37 +38,28 @@ def __init__(self, *args):
 
         self.framework.observe(self.on.install, self._on_install)
         self.framework.observe(self.on.config_changed, self._on_config_changed)
-        self.framework.observe(
-            self.on.mongodb_relation_joined, self._mongodb_relation_joined)
+        self.framework.observe(self.on.mongodb_relation_joined, self._mongodb_relation_joined)
         self.framework.observe(self.on.start, self._on_start)
         self.framework.observe(self.on.update_status, self._on_update_status)
-        self.framework.observe(self.on.get_primary_action,
-                               self._on_get_primary_action)
+        self.framework.observe(self.on.get_primary_action, self._on_get_primary_action)
 
     def _mongodb_relation_joined(self, event) -> None:
         if not self.unit.is_leader():
             return
 
         # do not initialise replica set until all peers have joined
-        if not self._number_of_expected_units == self._number_of_current_units:
-            logger.debug(
-                "planned units does not match joined units: %s != %s, will "
-                + "wait to initialise replica set until all planned units "
-                + "have joined",
-                str(self._number_of_expected_units),
-                str(self._number_of_current_units),
-            )
+        if not self._check_unit_count:
+            logger.debug("waiting to initialise replica set until all planned units have joined")
             return
```

Comment:
I don't recall if there is some nuance here - but in the next section you defer the event before returning - should you be deferring here?

Reply:
Yeah, this was something that _really_ confused me for awhile. If we defer here the **same** event will get called for the unit that created the `RelationJoinedEvent` meaning that if planned units will still not match the joined units. (this is because the other `RelationJoinedEvents` for the other units will not be called, and thus this condition will never be true and will forever defer.)

Reply:
Here is a longer explaination:

That's what I initially thought too! But actually having a defer here leads to a forever wait. For example:

Say you have 3 units (`A`,`B`,`C`) (and `A` is the leader). When the leader observes the first `relation-joined` event this is because **one** of the units `B` joined, **but `C` has not joined** thus:

`self.app.planned_units() == len(self._peers.units) + 1` = `3 = 2` 

so the condition is not satisfied. If we defer we will **keep waiting**, and we will **not** observe `C` joining. In order to observe `C` we will need to wait for the **second** `relation-joined` event. Once the second `relation-joined` even fires we will observe `C` and: 

`self.app.planned_units() == len(self._peers.units) + 1` = `3 = 3`

And then will will be ready.

TL;DR: We cannot defer since we need to allow all `relation-joined` events to fire so we can observe all the peers. 

I hope that makes sense 😅 its kind of confusing, I'm also super happy to talk about this on MM. This was something that tripped me up for _awhile_

Reply:
Nice, sounds good to me!